### ISSUE TITLE
Memory Leak on Windows when running example?

### ISSUE THREAD

**Issue by DhruvDh** [nannou-org/nannou#676]
When running example generative_designs/p_2_2_5_01.rs on Windows, if I resize the window the memory used on the GPU keeps going as the resize is in progress.

I just cloned the repository and ran the example like so -

```shell
C:\Users\dhruv\Documents\Repositories\nannou\(master)> cargo run --release --example p_2_2_5_01
    Finished release [optimized] target(s) in 0.66s
     Running `target\release\examples\p_2_2_5_01.exe`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AllocationError(OutOfMemory(Device))', C:\Users\dhruv\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-core-0.5.6\src\device\mod.rs:417:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\release\examples\p_2_2_5_01.exe` (exit code: 101)
```

If I resize slowly, the memory grows slowly, and vice versa. I know this by looking at the memory used graph in task manager.

**Comment by DhruvDh** [comment]
This only happens after a certain number of circles are drawn on screen. Also here is a full backtrace -

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AllocationError(OutOfMemory(Device))', C:\Users\dhruv\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-core-0.5.6\src\device\mod.rs:417:14
stack backtrace:
   0:     0x7ff7921ecb9e - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:108
   1:     0x7ff7921ecb9e - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff7921ecb9e - std::sys_common::backtrace::_print_fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff7921ecb9e - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff7922024db - core::fmt::write
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\fmt\mod.rs:1078
   5:     0x7ff7921e9828 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\io\mod.rs:1518
   6:     0x7ff7921ef98d - std::sys_common::backtrace::_print
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff7921ef98d - std::sys_common::backtrace::print
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff7921ef98d - std::panicking::default_hook::{{closure}}
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:208
   9:     0x7ff7921ef568 - std::panicking::default_hook
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:227
  10:     0x7ff7921f02ff - std::panicking::rust_panic_with_hook
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:593
  11:     0x7ff7921efe01 - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:499
  12:     0x7ff7921ed50f - std::sys_common::backtrace::__rust_end_short_backtrace<closure-0,!>
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff7921efd59 - std::panicking::begin_panic_handler
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:495
  14:     0x7ff7921fffb0 - core::panicking::panic_fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\panicking.rs:92
  15:     0x7ff7921ffdf3 - core::option::expect_none_failed
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\option.rs:1268
  16:     0x7ff7920bb20b - wgpu_core::device::Device<B>::create_texture::h28d057bc10a8a410
  17:     0x7ff792061691 - wgpu_core::device::<impl wgpu_core::hub::Global<G>>::device_create_texture::ha7a653fc6a3a31b6
  18:     0x7ff791fde137 - wgpu::Device::create_texture::h70b31b4d9c090452
  19:     0x7ff791f43035 - nannou::frame::RenderData::new::h5df8f70feea36e11
  20:     0x7ff791f16dc1 - nannou::window::Window::rebuild_swap_chain::h30ac11b0263c181d
  21:     0x7ff791f024eb - winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::{{closure}}::hde28dbb061918fb6
  22:     0x7ff791f28121 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h92b09291ff7cebd8
  23:     0x7ff791f487a5 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::register_window::h883344d5128c90ad
  24:     0x7ff791f47494 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::send_event::hf6bd23b3ec5eeb74
  25:     0x7ff791f2a5db - winit::platform_impl::platform::event_loop::public_window_callback::{{closure}}::h4cb002c124dbe823
  26:     0x7ff791f479d8 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::h1b2c75c0b32fea83
  27:     0x7ff791f75541 - winit::platform_impl::platform::event_loop::public_window_callback::h1aa2a4ccbd4ca78a
  28:     0x7ffc7251d3b6 - DefSubclassProc
  29:     0x7ffc7251d26c - DefSubclassProc
  30:     0x7ffc97b5e858 - CallWindowProcW
  31:     0x7ffc97b5e4ee - CallWindowProcW
  32:     0x7ffc46a5f190 - glPushClientAttrib
  33:     0x7ffc97b5e858 - CallWindowProcW
  34:     0x7ffc97b5e3dc - DispatchMessageW
  35:     0x7ffc97b70bc3 - SendMessageTimeoutW
  36:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  37:     0x7ffc97a01124 - NtUserMessageCall
  38:     0x7ffc97b5df33 - SendMessageW
  39:     0x7ffc97b5cbc3 - GetWindowTextW
  40:     0x7ffc97b5c2af - IsIconic
  41:     0x7ffc945db8d9 - IsCompositionActive
  42:     0x7ffc945db2c1 - IsCompositionActive
  43:     0x7ffc97b5c7e3 - GetWindowTextW
  44:     0x7ffc97b5e858 - CallWindowProcW
  45:     0x7ffc97b5e4ee - CallWindowProcW
  46:     0x7ffc7251d3b6 - DefSubclassProc
  47:     0x7ffc7251d14d - DefSubclassProc
  48:     0x7ff791f479d8 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::h1b2c75c0b32fea83
  49:     0x7ff791f75541 - winit::platform_impl::platform::event_loop::public_window_callback::h1aa2a4ccbd4ca78a
  50:     0x7ffc7251d3b6 - DefSubclassProc
  51:     0x7ffc7251d26c - DefSubclassProc
  52:     0x7ffc97b5e858 - CallWindowProcW
  53:     0x7ffc97b5e4ee - CallWindowProcW
  54:     0x7ffc46a5f190 - glPushClientAttrib
  55:     0x7ffc97b5e858 - CallWindowProcW
  56:     0x7ffc97b5e3dc - DispatchMessageW
  57:     0x7ffc97b75c00 - LookupIconIdFromDirectoryEx
  58:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  59:     0x7ffc97a01124 - NtUserMessageCall
  60:     0x7ffc97b5caf5 - GetWindowTextW
  61:     0x7ffc97b5c2af - IsIconic
  62:     0x7ffc945d9772 - GetUserColorPreference
  63:     0x7ffc945f34d2 - Ordinal96
  64:     0x7ffc945db859 - IsCompositionActive
  65:     0x7ffc945db2c1 - IsCompositionActive
  66:     0x7ffc97b5c7e3 - GetWindowTextW
  67:     0x7ff791f479d8 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::h1b2c75c0b32fea83
  68:     0x7ff791f75541 - winit::platform_impl::platform::event_loop::public_window_callback::h1aa2a4ccbd4ca78a
  69:     0x7ffc7251d3b6 - DefSubclassProc
  70:     0x7ffc7251d26c - DefSubclassProc
  71:     0x7ffc97b5e858 - CallWindowProcW
  72:     0x7ffc97b5e4ee - CallWindowProcW
  73:     0x7ffc46a5f190 - glPushClientAttrib
  74:     0x7ffc97b5e858 - CallWindowProcW
  75:     0x7ffc97b5e3dc - DispatchMessageW
  76:     0x7ffc97b70bc3 - SendMessageTimeoutW
  77:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  78:     0x7ffc97a01124 - NtUserMessageCall
  79:     0x7ffc97b5caf5 - GetWindowTextW
  80:     0x7ffc97b5c2af - IsIconic
  81:     0x7ffc945d9772 - GetUserColorPreference
  82:     0x7ffc945f3427 - Ordinal96
  83:     0x7ffc945db859 - IsCompositionActive
  84:     0x7ffc945db2c1 - IsCompositionActive
  85:     0x7ffc97b5c7e3 - GetWindowTextW
  86:     0x7ffc97b5e858 - CallWindowProcW
  87:     0x7ffc97b5e4ee - CallWindowProcW
  88:     0x7ffc7251d3b6 - DefSubclassProc
  89:     0x7ffc7251d14d - DefSubclassProc
  90:     0x7ff791f479d8 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::h1b2c75c0b32fea83
  91:     0x7ff791f75541 - winit::platform_impl::platform::event_loop::public_window_callback::h1aa2a4ccbd4ca78a
  92:     0x7ffc7251d3b6 - DefSubclassProc
  93:     0x7ffc7251d26c - DefSubclassProc
  94:     0x7ffc97b5e858 - CallWindowProcW
  95:     0x7ffc97b5e4ee - CallWindowProcW
  96:     0x7ffc46a5f190 - glPushClientAttrib
  97:     0x7ffc97b5e858 - CallWindowProcW
  98:     0x7ffc97b5e299 - DispatchMessageW
  99:     0x7ff791f00cb4 - winit::platform_impl::platform::event_loop::EventLoop<T>::run::h5f9d04ad056d476f
 100:     0x7ff791ef57e2 - alloc::alloc::box_free::h8da73cd72efbcf0a
 101:     0x7ff791ef5f09 - nannou::app::Builder<M,E>::run::h5e31d89acba7b0a4
 102:     0x7ff791f03845 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::poll::h3fd08a37721c2b0e
 103:     0x7ff791f0c2d6 - std::sys_common::backtrace::__rust_end_short_backtrace::h96d18383a9b93335
 104:     0x7ff791f0c2ec - std::rt::lang_start::{{closure}}::h0629636048c376f5
 105:     0x7ff7921f0643 - core::ops::function::impls::{{impl}}::call_once
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\library\core\src\ops\function.rs:280
 106:     0x7ff7921f0643 - std::panicking::try::do_call
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:381
 107:     0x7ff7921f0643 - std::panicking::try
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:345
 108:     0x7ff7921f0643 - std::panic::catch_unwind
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panic.rs:396
 109:     0x7ff7921f0643 - std::rt::lang_start_internal
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\rt.rs:51
 110:     0x7ff791f04697 - main
 111:     0x7ff7922af9c4 - invoke_main
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
 112:     0x7ff7922af9c4 - __scrt_common_main_seh
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
 113:     0x7ffc99477034 - BaseThreadInitThunk
 114:     0x7ffc99a1cec1 - RtlUserThreadStart
error: process didn't exit successfully: `target\release\examples\p_2_2_5_01.exe` (exit code: 101)
```

**Comment by mitchmindtree** [comment]
Thanks for the issue and digging into this!

This sounds suspiciously similar to #491 

Do you have the same issue when using the branch at #665? I'm curious if the new wgpu version fixes anything.

If not, we're almost certainly doing something wrong somewhere.

**Comment by DhruvDh** [comment]
Oh, I am sorry I forgot to search for similar issues before filing one. I am new to this.

The same thing happens when using the branch at #665, here's another backtrace -

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AllocationError(OutOfMemory(Device))', C:\Users\dhruv\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-core-0.5.6\src\device\mod.rs:417:14
stack backtrace:
   0:     0x7ff6b04cc8de - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:108
   1:     0x7ff6b04cc8de - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff6b04cc8de - std::sys_common::backtrace::_print_fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff6b04cc8de - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff6b04e221b - core::fmt::write
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\fmt\mod.rs:1078
   5:     0x7ff6b04c9568 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\io\mod.rs:1518
   6:     0x7ff6b04cf6cd - std::sys_common::backtrace::_print
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff6b04cf6cd - std::sys_common::backtrace::print
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff6b04cf6cd - std::panicking::default_hook::{{closure}}
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:208
   9:     0x7ff6b04cf2a8 - std::panicking::default_hook
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:227
  10:     0x7ff6b04d003f - std::panicking::rust_panic_with_hook
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:593
  11:     0x7ff6b04cfb41 - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:499
  12:     0x7ff6b04cd24f - std::sys_common::backtrace::__rust_end_short_backtrace<closure-0,!>
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff6b04cfa99 - std::panicking::begin_panic_handler
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:495
  14:     0x7ff6b04dfcf0 - core::panicking::panic_fmt
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\panicking.rs:92
  15:     0x7ff6b04dfb33 - core::option::expect_none_failed
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\core\src\option.rs:1268
  16:     0x7ff6b039a1cb - wgpu_core::device::Device<B>::create_texture::h126d975fc230fa54
  17:     0x7ff6b0341851 - wgpu_core::device::<impl wgpu_core::hub::Global<G>>::device_create_texture::h40345a3b5bbe3090
  18:     0x7ff6b02bd697 - wgpu::Device::create_texture::h9f243f1c4d8923bc
  19:     0x7ff6b0221d95 - nannou::frame::RenderData::new::ha109a4592aa10b8c
  20:     0x7ff6b01f6b31 - nannou::window::Window::rebuild_swap_chain::hc39a060de7582c15
  21:     0x7ff6b01e24eb - winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::{{closure}}::h07588408bd8d8774
  22:     0x7ff6b0207f91 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hc50cac079d116775
  23:     0x7ff6b0227505 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::register_window::h9f0dc5918cc9b4bc
  24:     0x7ff6b02261f4 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::send_event::h049bd380ff0ee103
  25:     0x7ff6b0209e96 - winit::platform_impl::platform::event_loop::public_window_callback::{{closure}}::h22a601d6eeb65b47
  26:     0x7ff6b0226738 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::hbedea8d98a661657
  27:     0x7ff6b0255681 - winit::platform_impl::platform::event_loop::public_window_callback::ha3d6090819a42343
  28:     0x7ffc7251d3b6 - DefSubclassProc
  29:     0x7ffc7251d26c - DefSubclassProc
  30:     0x7ffc97b5e858 - CallWindowProcW
  31:     0x7ffc97b5e4ee - CallWindowProcW
  32:     0x7ffc6aeaf190 - glPushClientAttrib
  33:     0x7ffc97b5e858 - CallWindowProcW
  34:     0x7ffc97b5e3dc - DispatchMessageW
  35:     0x7ffc97b70bc3 - SendMessageTimeoutW
  36:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  37:     0x7ffc97a01124 - NtUserMessageCall
  38:     0x7ffc97b5df33 - SendMessageW
  39:     0x7ffc97b5cbc3 - GetWindowTextW
  40:     0x7ffc97b5c2af - IsIconic
  41:     0x7ffc945db8d9 - IsCompositionActive
  42:     0x7ffc945db2c1 - IsCompositionActive
  43:     0x7ffc97b5c7e3 - GetWindowTextW
  44:     0x7ffc97b5e858 - CallWindowProcW
  45:     0x7ffc97b5e4ee - CallWindowProcW
  46:     0x7ffc7251d3b6 - DefSubclassProc
  47:     0x7ffc7251d14d - DefSubclassProc
  48:     0x7ff6b0226738 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::hbedea8d98a661657
  49:     0x7ff6b0255681 - winit::platform_impl::platform::event_loop::public_window_callback::ha3d6090819a42343
  50:     0x7ffc7251d3b6 - DefSubclassProc
  51:     0x7ffc7251d26c - DefSubclassProc
  52:     0x7ffc97b5e858 - CallWindowProcW
  53:     0x7ffc97b5e4ee - CallWindowProcW
  54:     0x7ffc6aeaf190 - glPushClientAttrib
  55:     0x7ffc97b5e858 - CallWindowProcW
  56:     0x7ffc97b5e3dc - DispatchMessageW
  57:     0x7ffc97b75c00 - LookupIconIdFromDirectoryEx
  58:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  59:     0x7ffc97a01124 - NtUserMessageCall
  60:     0x7ffc97b5caf5 - GetWindowTextW
  61:     0x7ffc97b5c2af - IsIconic
  62:     0x7ffc945d9772 - GetUserColorPreference
  63:     0x7ffc945f34d2 - Ordinal96
  64:     0x7ffc945db859 - IsCompositionActive
  65:     0x7ffc945db2c1 - IsCompositionActive
  66:     0x7ffc97b5c7e3 - GetWindowTextW
  67:     0x7ff6b0226738 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::hbedea8d98a661657
  68:     0x7ff6b0255681 - winit::platform_impl::platform::event_loop::public_window_callback::ha3d6090819a42343
  69:     0x7ffc7251d3b6 - DefSubclassProc
  70:     0x7ffc7251d26c - DefSubclassProc
  71:     0x7ffc97b5e858 - CallWindowProcW
  72:     0x7ffc97b5e4ee - CallWindowProcW
  73:     0x7ffc6aeaf190 - glPushClientAttrib
  74:     0x7ffc97b5e858 - CallWindowProcW
  75:     0x7ffc97b5e3dc - DispatchMessageW
  76:     0x7ffc97b70bc3 - SendMessageTimeoutW
  77:     0x7ffc99a6fbd4 - KiUserCallbackDispatcher
  78:     0x7ffc97a01124 - NtUserMessageCall
  79:     0x7ffc97b5caf5 - GetWindowTextW
  80:     0x7ffc97b5c2af - IsIconic
  81:     0x7ffc945d9772 - GetUserColorPreference
  82:     0x7ffc945f3427 - Ordinal96
  83:     0x7ffc945db859 - IsCompositionActive
  84:     0x7ffc945db2c1 - IsCompositionActive
  85:     0x7ffc97b5c7e3 - GetWindowTextW
  86:     0x7ffc97b5e858 - CallWindowProcW
  87:     0x7ffc97b5e4ee - CallWindowProcW
  88:     0x7ffc7251d3b6 - DefSubclassProc
  89:     0x7ffc7251d14d - DefSubclassProc
  90:     0x7ff6b0226738 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::catch_unwind::hbedea8d98a661657
  91:     0x7ff6b0255681 - winit::platform_impl::platform::event_loop::public_window_callback::ha3d6090819a42343
  92:     0x7ffc7251d3b6 - DefSubclassProc
  93:     0x7ffc7251d26c - DefSubclassProc
  94:     0x7ffc97b5e858 - CallWindowProcW
  95:     0x7ffc97b5e4ee - CallWindowProcW
  96:     0x7ffc6aeaf190 - glPushClientAttrib
  97:     0x7ffc97b5e858 - CallWindowProcW
  98:     0x7ffc97b5e299 - DispatchMessageW
  99:     0x7ff6b01e0cb4 - winit::platform_impl::platform::event_loop::EventLoop<T>::run::h7348ece4b4144699
 100:     0x7ff6b01d57e2 - alloc::alloc::box_free::h01e49860b15f036b
 101:     0x7ff6b01d5f09 - nannou::app::Builder<M,E>::run::hbd0fa09222900d7e
 102:     0x7ff6b01e3845 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<T>::poll::h9d0b6952fcf907fd
 103:     0x7ff6b01ec046 - std::sys_common::backtrace::__rust_end_short_backtrace::h96d18383a9b93335
 104:     0x7ff6b01ec05c - std::rt::lang_start::{{closure}}::h0629636048c376f5
 105:     0x7ff6b04d0383 - core::ops::function::impls::{{impl}}::call_once
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\library\core\src\ops\function.rs:280
 106:     0x7ff6b04d0383 - std::panicking::try::do_call
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:381
 107:     0x7ff6b04d0383 - std::panicking::try
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panicking.rs:345
 108:     0x7ff6b04d0383 - std::panic::catch_unwind
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\panic.rs:396
 109:     0x7ff6b04d0383 - std::rt::lang_start_internal
                               at /rustc/5404efc28a0cddee103ef6396c48ea71ff9631c8\/library\std\src\rt.rs:51
 110:     0x7ff6b01e4697 - main
 111:     0x7ff6b058f704 - invoke_main
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
 112:     0x7ff6b058f704 - __scrt_common_main_seh
                               at d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
 113:     0x7ffc99477034 - BaseThreadInitThunk
 114:     0x7ffc99a1cec1 - RtlUserThreadStart
error: process didn't exit successfully: `target\release\examples\p_2_2_5_01.exe` (exit code: 101)
```

**Comment by MacTuitui** [comment]
I'm hitting this as well (linux, though, I haven't run into the same issue on my mac, so probably vulkan related). My intuition is that the geometry and textures (well, I guess all the buffers) are not freed once they've been used on the GPU, so it's probably a wgpu issue. 

On my latest series, I can draw a hundred frames with rects and then OOM, but when I switch to ellipses I'm lucky if I get a few frames. 

Now how will I manage today's one...

**Comment by MacTuitui** [comment]
Now for my use case (lots of geometry), using the wgpu-0.6-3 branch works.

(Luckily I'm not using straight wgpu these days, otherwise everything would have broke!)

**Comment by anlumo** [comment]
(I got asked to reply to this here. Note that I have no idea about nannou.)

In our wgpu-using program, we've seen something similar. It's not a memory leak, the memory management is deferred to a later point, and that point is after the GPU runs out of memory due to the frequent reallocations of the buffers, crashing the rendering context.
There was an attempt at fixing this in wgpu, but my graphics programmer told me that it's not really solving the issue completely.

**Comment by danwilhelm** [comment]
I discovered the cause of this issue and submitted it as part of #720. I initially also thought it was likely the wgpu memory management, as @anlumo suggested. But, I ported the relevant code to wgpu 0.7 to no avail. (wgpu 0.7 attempts to immediately deallocate textures, instead of  deferring removal.)

The cause ended up being a memory leak in nannou:
`TextureId(Arc::into_raw(self.handle.clone()) as usize)` 
-> `TextureId(Arc::as_ptr(&self.handle) as usize)`

The `Arc` reference count was incremented via `clone`, then promptly forgotten about via `Arc::into_raw`. This caused the linear framebuffer textures never to be dropped/deallocated, e.g. upon window resize.
